src/ipc/Queue.h

```diff
@@ -410,6 +425,92 @@ OneToOneUniQueue::push(const Value &value, QueueReader *const reader)
     return wasEmpty && (!reader || reader->raiseSignal());
 }
 
+template <class Value>
+void
+OneToOneUniQueue::statIn(std::ostream &os, const int localProcessId, const int remoteProcessId) const
+{
+    os << "  kid" << localProcessId << " receiving from kid" << remoteProcessId << ": ";
+    // Nobody can modify our theOut so, after capturing some valid theSize value
+    // in count, we can reliably report all [theOut, theOut+count) items that
+    // were queued at theSize capturing time. We will miss new items push()ed by
+    // the other side, but it is OK -- we report state at the capturing time.
+    const auto count = theSize.load();
+    statOpen(os, "other", "popIndex", count);
+    statSamples<Value>(os, theOut, count);
+    statClose(os);
+}
+
+template <class Value>
+void
+OneToOneUniQueue::statOut(std::ostream &os, const int localProcessId, const int remoteProcessId) const
+{
+    os << "  kid" << localProcessId << " sending to kid" << remoteProcessId << ": ";
+    // Nobody can modify our theIn so, after capturing some valid theSize value
+    // in count, we can reliably report all [theIn-count, theIn) items that were
+    // queued at theSize capturing time. We may report items already pop()ed by
+    // the other side, but that is OK because pop() does not modify items -- it
+    // only increments theOut.
+    const auto count = theSize.load();
+    statOpen(os, "pushIndex", "other", count);
+    statSamples<Value>(os, theIn - count, count); // unsigned offset underflow OK
+    statClose(os);
+}
+
+/// report a sample of [start, start + size) items
+template <class Value>
+void
+OneToOneUniQueue::statSamples(std::ostream &os, const unsigned int start, const uint32_t count) const
+{
+    if (!count) {
+        os << " ";
+        return;
+    }
+
+    os << ", items: [\n";
+    // report a few leading and trailing items, without repetitions
+    const auto sampleSize = std::min(3U, count); // leading (and max) sample
+    statRange<Value>(os, start, 0, sampleSize);
+    if (sampleSize < count) { // the first sample did not show some items
+        // The `start` offset aside, the first sample reported all items
+        // below the sampleSize offset. The second sample needs to report
+        // the last sampleSize items (i.e. starting at count-sampleSize
+        // offset) except those already reported by the first sample.
+        const auto secondSampleOffset = std::max(sampleSize, count - sampleSize);
+        const auto secondSampleSize = std::min(sampleSize, count - sampleSize);
+
+        // but first we print a sample separator, unless there are no items
+        // between the samples or the separator hides the only unsampled item
+        const auto bothSamples = sampleSize + secondSampleSize;
+        if (bothSamples + 1U == count)
+            statRange<Value>(os, start, sampleSize, 1);
+        else if (count > bothSamples)
+            os << "    # ... " << (count - bothSamples) << " items not shown ...\n";
```

Comment:
Samples are useful when diagnosing some problems. For example, they can tell humans whether the whole queue got stuck for a long time, whether two kids have a compatible/expected view of the queue that links them, and whether some queues are overflown while showing messages that Squid already reported as timed out.

Reply:
So this is a debug output, not a management report ?

Reply:
Not sure what does the 'management report' mean in this context, but IIRC there were there options:
1. Show no samples at all.
2. Show all samples.
3. Show only a few samples.

We selected (3) as a compromise.

Reply:
> So this is a debug output, not a management report ?

I cannot answer this question because I do not know what you mean by "debug output" and "management report". You know what "this" is because you see the code and the [output](https://github.com/squid-cache/squid/pull/690#issuecomment-653958250) it produces, so I suspect this is some kind of an indirect question designed to expose some perceived flaw. Please consider rephrasing the question in a more direct/specific way.

FWIW, I think there is nothing wrong with adding samples to cache manager reports. I personally find them useful and sometimes even insightful when looking at these reports during bug triage. Moreover, showing thousands of entries (by default) has been proven to be a _bad_ idea because generating the corresponding report usually blocks Squid for a long time, overflows buffers, and leads to other problems/[bugs](https://github.com/squid-cache/squid/pull/536). If somebody wants a full queue snapshot, they can enhance this code to produce one when explicitly requested via a cache manager request parameter.

Reply:
I agree there is nothing wrong with reporting this data. Just trying to get an understanding of the use-cases here.

FWIW; I meant "debug output" for reports only used when debugging or troubleshooting a problem. Which would naturally have a much, much lower frequency of use than reports used routinely to see what the traffic is doing (eg info report) - that later being the "management report"'s.

Since this appears to be in the former debug group I wondered why the need to compromise and show so few entries. That is an output more suited to summaries for management reports. Debug would seem to be more useful getting the full data.  From your responses I understand that only the beginning and ending of the queue is useful for the debug.

Reply:
Thank you for resolving this change request. While I disagree with your analysis and do not recommend using the proposed terminology, I do not think we need to reach consensus on the anticipated reporting frequency or the terminology used to describe the purpose(s) of this cache manager report. We can revisit the terminology issue later if such a need arises, of course.